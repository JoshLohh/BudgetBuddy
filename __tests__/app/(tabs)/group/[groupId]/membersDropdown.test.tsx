/**
 * Test suite for MembersDropdown component
 * Covers rendering, interaction, routing, and callbacks.
 * Uses mocks for Expo Router and verifies UI behavior on toggle, navigation, removal, and adding member.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import MembersDropdown from '@/app/(tabs)/group/[groupId]/membersDropdown';
import { useRouter } from 'expo-router';

// Mock Expo Router
jest.mock('expo-router', () => ({
  useRouter: jest.fn(),
}));

describe('MembersDropdown', () => {
  const mockPush = jest.fn();
  const currentUserId = 'user1';
  const group = { $id: 'group1', title: 'Test Group', members: ['user1', 'user2'], createdBy: 'user1' };
  const memberProfiles = [
    { userId: 'user1', username: 'Alice', avatar: 'avatar1.png' },
    { userId: 'user2', username: 'Bob', avatar: 'avatar2.png' },
  ];
  const emptyMemberProfiles: typeof memberProfiles = [];

  beforeEach(() => {
    // Reset mocks & provide router mock
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  });

  it('renders header and toggle button with correct initial state', () => {
    const { getByText } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={false}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    );
    expect(getByText('Group Members')).toBeTruthy();
    expect(getByText('Show Members')).toBeTruthy();
  });

  it('toggles membersExpanded when toggle button is pressed', () => {
    let expanded = false;
    const setMembersExpanded = jest.fn((fn) => {
      expanded = fn(expanded);
    });

    const { getByText, rerender } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={expanded}
        setMembersExpanded={setMembersExpanded}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    );

    const toggleButton = getByText('Show Members');
    fireEvent.press(toggleButton);

    expect(setMembersExpanded).toHaveBeenCalled();

    // Rerender with updated state
    rerender(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={expanded}
        setMembersExpanded={setMembersExpanded}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    );

    if (expanded) {
      expect(getByText('Hide Members')).toBeTruthy();
    }
  });

  it('shows "No members." message when memberProfiles is empty and expanded', () => {
    const { getByText } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={emptyMemberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    );

    expect(getByText('No members.')).toBeTruthy();
  });

  it('renders member list with remove button for other members', () => {
    const handleRemoveMember = jest.fn();
    const { getByTestId, queryByTestId } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={handleRemoveMember}
        setSearchModalVisible={() => {}}
      />
    );
  
    // Current user should NOT have remove button
    expect(queryByTestId('remove-btn-user1')).toBeNull();
  
    // Other member should have remove button
    const removeButton = getByTestId('remove-btn-user2');
    expect(removeButton).toBeTruthy();
  
    fireEvent.press(removeButton);
    expect(handleRemoveMember).toHaveBeenCalledWith('user2');
  });
  
  it('navigates to user profile when member pressed', () => {
    const { getByTestId } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    );
  
    const memberTouchable = getByTestId('member-user2');
    fireEvent.press(memberTouchable);
  
    expect(mockPush).toHaveBeenCalledWith({
      pathname: '/user/[userId]',
      params: { userId: 'user2', groupId: 'group1' },
    });
  });
  

  it('calls setSearchModalVisible when Add Member button pressed', () => {
    const setSearchModalVisible = jest.fn();

    const { getByText } = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={setSearchModalVisible}
      />
    );

    const addButton = getByText('+ Add Member');
    fireEvent.press(addButton);

    expect(setSearchModalVisible).toHaveBeenCalledWith(true);
  });

  //Snapshots
  it('matches snapshot when collapsed', () => {
    const tree = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={false}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when expanded with members', () => {
    const tree = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={memberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when expanded with no members', () => {
    const tree = render(
      <MembersDropdown
        currentUserId={currentUserId}
        group={group}
        membersExpanded={true}
        setMembersExpanded={() => {}}
        memberProfiles={emptyMemberProfiles}
        handleRemoveMember={() => {}}
        setSearchModalVisible={() => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
