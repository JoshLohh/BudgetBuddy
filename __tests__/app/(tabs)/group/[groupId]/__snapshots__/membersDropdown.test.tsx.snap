// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MembersDropdown matches snapshot when collapsed 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 17,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Group Members
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#1e88e5",
              "fontWeight": "bold",
            },
          ]
        }
      >
        Show
         Members
      </Text>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#1e88e5",
              "fontSize": 18,
            },
            {
              "marginLeft": 4,
            },
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`MembersDropdown matches snapshot when expanded with members 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 17,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Group Members
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#1e88e5",
              "fontWeight": "bold",
            },
          ]
        }
      >
        Hide
         Members
      </Text>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#1e88e5",
              "fontSize": 18,
            },
            {
              "marginLeft": 4,
            },
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 4,
        "marginTop": 6,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingRight": 4,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginRight": 0,
            "opacity": 1,
          }
        }
        testID="member-user1"
      >
        <Image
          source={
            {
              "uri": "avatar1.png",
            }
          }
          style={
            {
              "backgroundColor": "#eee",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "flex": 1,
              "marginLeft": 12,
            },
          ]
        }
      >
        Alice
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingRight": 4,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginRight": 0,
            "opacity": 1,
          }
        }
        testID="member-user2"
      >
        <Image
          source={
            {
              "uri": "avatar2.png",
            }
          }
          style={
            {
              "backgroundColor": "#eee",
              "borderRadius": 18,
              "height": 36,
              "width": 36,
            }
          }
        />
      </View>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "flex": 1,
              "marginLeft": 12,
            },
          ]
        }
      >
        Bob
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "flex-end",
            "marginLeft": 8,
            "opacity": 1,
            "padding": 2,
          }
        }
        testID="remove-btn-user2"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#e57373",
                "fontSize": 20,
              },
              undefined,
              {
                "fontFamily": "ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderRadius": 6,
              "marginVertical": 10,
              "padding": 18,
            },
            {
              "backgroundColor": "#0a7ea4",
            },
            false,
            {
              "alignSelf": "flex-end",
              "backgroundColor": "#005c99",
              "justifyContent": "center",
              "marginTop": 8,
              "width": 150,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#11181C",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              {
                "color": "#fff",
                "textAlign": "center",
              },
            ]
          }
        >
          + Add Member
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`MembersDropdown matches snapshot when expanded with no members 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 17,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Group Members
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#1e88e5",
              "fontWeight": "bold",
            },
          ]
        }
      >
        Hide
         Members
      </Text>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#1e88e5",
              "fontSize": 18,
            },
            {
              "marginLeft": 4,
            },
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "marginBottom": 4,
        "marginTop": 6,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "color": "#aaa",
            "marginLeft": 4,
          },
        ]
      }
    >
      No members.
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderRadius": 6,
              "marginVertical": 10,
              "padding": 18,
            },
            {
              "backgroundColor": "#0a7ea4",
            },
            false,
            {
              "alignSelf": "flex-end",
              "backgroundColor": "#005c99",
              "justifyContent": "center",
              "marginTop": 8,
              "width": 150,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#11181C",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              {
                "color": "#fff",
                "textAlign": "center",
              },
            ]
          }
        >
          + Add Member
        </Text>
      </View>
    </View>
  </View>
</View>
`;
